test/python/test_ibmqjob.py

```diff
@@ -0,0 +1,143 @@
+# =============================================================================
+
+"""IBMQJob Test."""
+
+import unittest
+import time
+import numpy
+from scipy.stats import chi2_contingency
+
+from qiskit import (ClassicalRegister, QuantumCircuit, QuantumRegister,
+                    QuantumJob)
+import qiskit._compiler
+from qiskit.backends.ibmq import IBMQProvider
+from .common import requires_qe_access, QiskitTestCase, slow_test
+
+
+class TestIBMQJob(QiskitTestCase):
+    """
+    Test ibmqjob module.
+    """
+
+    @classmethod
+    @requires_qe_access
+    def setUpClass(cls, QE_TOKEN, QE_URL):
+        # pylint: disable=arguments-differ
+        super().setUpClass()
+        # create QuantumCircuit
+        qr = QuantumRegister(2, 'q')
+        cr = ClassicalRegister(2, 'c')
+        qc = QuantumCircuit(qr, cr)
+        qc.h(qr[0])
+        qc.cx(qr[0], qr[1])
+        qc.measure(qr, cr)
+        cls._qc = qc
+        cls._provider = IBMQProvider(QE_TOKEN, QE_URL)
+
+    def test_run(self):
+        backend = self._provider.get_backend('ibmqx_qasm_simulator')
+        qobj = qiskit._compiler.compile(self._qc, backend)
+        quantum_job = QuantumJob(qobj, backend, shots=1024, preformatted=True)
+        job = backend.run(quantum_job)
+        result = job.result()
+        counts_qx = result.get_counts(result.get_names()[0])
+        counts_ex = {'00': 512, '11': 512}
+        states = counts_qx.keys() | counts_ex.keys()
+        # contingency table
+        ctable = numpy.array([[counts_qx.get(key, 0) for key in states],
+                              [counts_ex.get(key, 0) for key in states]])
+        contingency = chi2_contingency(ctable)
+        self.log.info('chi2_contingency: %s', str(contingency))
+        self.assertGreater(contingency[1], 0.01)
+
+    @slow_test
+    def test_run_async_simulator(self):
+        backend = self._provider.available_backends({'simulator': True})[0]
+        self.log.info('submitting to backend %s', backend.name)
+        num_qubits = 5
+        qr = QuantumRegister(num_qubits, 'qr')
+        cr = ClassicalRegister(num_qubits, 'cr')
+        qc = QuantumCircuit(qr, cr)
+        for i in range(num_qubits-1):
+            qc.cx(qr[i], qr[i+1])
+        qc.measure(qr, cr)
+        qobj = qiskit._compiler.compile([qc]*10, backend)
+        quantum_job = QuantumJob(qobj, backend, shots=1e3, preformatted=True)
+        num_jobs = 5
+        job_array = [backend.run(quantum_job) for _ in range(num_jobs)]
+        time.sleep(0.1)  # give time for jobs to start (better way?)
+        num_queued = sum([job.queued for job in job_array])
+        num_running = sum([job.running for job in job_array])
+        self.log.info('number of currently queued jobs: %d/%d',
+                      num_queued, num_jobs)
+        self.log.info('number of currently running jobs: %d/%d',
+                      num_running, num_jobs)
+        self.assertTrue(all([(job.running or job.queued) for job in job_array]))
```

Comment:
Can you replace this line with an slightly expanded version? Something along the lines of:

```
        self.assertTrue(all([(job.running or job.queued or job.done) for job in job_array]))

        # Ensure job ids are unique.
        job_ids = [job.job_id for job in job_array]
        self.assertEqual(sorted(job_ids), sorted(list(set(job_ids))))

        # Wait for all the results.
        result_array = [job.result() for job in job_array]

        # Ensure all jobs have finished.
        self.assertTrue(all([job.done for job in job_array]))
        self.assertTrue(all([result.get_status() == 'COMPLETED' for result in result_array]))
```

The `or job.done`, along with the fix mentioned by @ajavadia at https://github.com/QISKit/qiskit-sdk-py/pull/403/files#r186906393, should be enough for passing the test in its current state - there is the minor issue that the line would make three queries to the API (one for each `job.something`) and increase the running time, but seems we can afford that for the moment.

The rest of the lines are for making the tests more complete: not only launching the jobs, but ensuring that they get completed and that they handle different job_ids. For the simulator, it does seem indeed that all the jobs will be in `job.done` state by the time they reach line 91 (but it seems fine too).